### DISCUSSION TITLE
Value types

### DISCUSSION THREAD

**Discussion by jlkiri** [kdl-org/kdl#3] [General]
I wonder what is the possible real-world use of mixed value types like:

```
foo 1 2 3 "hello"
```

I'm not sure yet but wouldn't it be a pain to deserialize this in languages where an array of values can only have one type?

**Comment by zkat** [comment]
I think this would work out fine when working with, say, serde, because it would generate errors if it doesn't get the general structure that it expects to be deserializing? Much like serde_json supports weird stuff like "string OR object" in values?

**Reply by zkat** [reply]
it does make me think that any serde bindings to this would need to be able to handle both the "array of things" case and the "first arg maps to this, second arg maps to that" for heterogenous node values.

Something like:
```rust
struct Foo {
    #[serde(rename = "$0")]
    pub name: String,
    #[serde(rename = "$1")]
    pub age: usize,
}
```

which would then apply to `foo "Alex" 33`

**Comment by xpe** [comment]
> I'm not sure yet but wouldn't it be a pain to deserialize this in languages where an array of values can only have one type?

Depending on the language, maybe/perhaps.

In Rust, where arrays must be homogenous, one could:

- use an array of an enum type
- like @zkat says in the other answer, one could deserialize to a structure
- deserialize to a tuple type such as `(String, usize)`

**Comment by tabatkins** [comment]
In general, the answer to both "what's the real world use-case" and "what do you do in certain languages" is "the exact same as for `[1, 2, 3, "foo"]` in JSON". I don't believe there's anything unique to KDL here, it's just a matter of possible data-representation mismatches, which applies between any two data languages.
